rs/rust_canisters/random_traffic_test/src/lib.rs

```diff
@@ -81,48 +83,57 @@ impl Config {
 pub enum Reply {
     /// A response including a data payload of a distinct size was received.
     Bytes(u32),
-    /// The call was rejected synchronously by the system with an error code.
-    SynchronousRejection(i32),
-    /// The call was rejected asynchronously through a reject response, which
-    /// includes an error code and a reject message.
-    AsynchronousRejection(i32, String),
+    /// The call was rejected with a reject code and a reject message.
+    Reject(u32, String),
 }
 
-/// Record for one outgoing call. Records how many bytes were sent out; and what kind of
-/// reply was received (either data or a synchronous or asynchronous rejection).
+/// Record for an outgoing call.
 #[derive(Serialize, Deserialize, Clone, PartialEq, Eq, CandidType)]
 pub struct Record {
+    /// The `receiver` of the call.
     pub receiver: CanisterId,
+    /// The caller if any, i.e. if this is a downstream call to a call made by `caller`.
+    pub caller: Option<CanisterId>,
```

Comment:
I guess the call tree ID (the ID given to a call spawned in a heartbeat and then passed on to downstream calls) and the call depth would be enough to reconstruct a full call tree from the records. But I thought it still useful to know who the caller was. I suppose it would make such an algorithm more efficient but it's also much easier to follow when manually reading the records.